src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py

```diff
@@ -470,16 +470,7 @@ def __getPilotToken(self):
 
         :return: S_OK/S_ERROR, Token object as Value
         """
-
-        result = Registry.getUsernameForDN(self.pilotDN)
-        if not result["OK"]:
-            return result
-        username = result["Value"]
-        result = gTokenManager.getToken(
-            username=username,
-            userGroup=self.pilotGroup,
-            requiredTimeLeft=3600,
-        )
+        result = gTokenManager.getToken(userGroup=self.pilotGroup, requiredTimeLeft=600)
```

Comment:
I think it is better to make client or user token to use to be configurable. E.g. GenericPilotUser = client should trigger getToken with userName=None (or "client") which is interpreted as using Client.

Reply:
Rather the other way around no ? If there is a username defined in the `SiteDirector` config, then use it, otherwise don't set it. But we should not have magic word like *client*

Reply:
I agree that a magic word sounds like a hack.
But I am not convinced about choosing what to use based on whether a username is defined in the `SiteDirector`.
We would have something like this in `SiteDirector.beginExecution()`:


```python

self.pilotUser = self.am_getOption("PilotUser", None)
# would need to change the logic in findGenericPilotCredentials() to potentially return a PilotUser
result = findGenericPilotCredentials(vo=self.vo, pilotDN=self.pilotDN, pilotUser=self.pilotUser, pilotGroup=self.pilotGroup) 
if not result["OK"]:
    return result
self.pilotUser, self.pilotDN, self.pilotGroup = result["Value"]
```

This seems tricky.
We would need both the `PilotDN` (for the proxy) and the `PilotUser` (for the token).
And they could be different.

Just relying on `PilotDN` and getting the `PilotUser` from it would not be a valid solution at the moment: we still need `PilotDN` to get a proxy for the pilots, so it has to be set in any case (which means we don't use the `client_credentials` flow).
We could also add a flag in the `SiteDirector` to know if we should use the username or not, but that seems inconvenient.

What would be the advantage of these solutions? Why would we need to submit pilots with a user access token? (at least for now we know that the client access tokens will be accepted by the sites).

Reply:
A summary of a discussion with @atsareg:

HTCondor sites want to know which token belongs to which VO.
The WLCG AuthZ group emits the idea that this could be extracted from the `issuer` claim by the sites.
VOs depending on Check In would have the same `issuer` claims so this would not be a good idea.

The WLCG AuthZ group suggested that they could have one instance but multiple issuer URLs (one per VO).
The Check In group did not accept this proposal as it would apparently go against the OIDC guidelines.
Another option for the sites, would be to use the `introspection` endpoint of the issuer to get more details about a given token.
This cannot be done for every job submission and would need to be done once and stored in a mapping file by the sites (this `issuer+subject` = this VO).
This is extra work to do for the sites, so there is no guarantee that they will accept.

They are working on a plugin to provide a common way to validate tokens.
This could be pushed at this level, but there are apparently no plans for that at the moment.
In any case, this needs some time to be developed.

This is why, at the moment, the SiteDirector should support both user and client access tokens.

Reply:
Where can we actually see the status of token deployment and discussions in EGI CheckIn ?  
WLCG has public info on that, with precise milestones and documented minutes, so it is something we can work with. But we are completely blind with respect to CheckIn, and we can't work properly on hearsay.

Reply:
Update on the EGI Check-In. The Check-In related [doc](https://docs.google.com/document/d/1q67BWeFBmFmQ0crG5bVWwqecVMdN8PROYnciHO5JhRo/edit#heading=h.ropcuc1skhp5). The main points:

Check-In implemented compute scopes in the dev service that we are testing now.

Check-In agrees to allow client credentials tokens to be used to access CEs. These tokens should be obtained with dedicated clients that can get compute scopes (pilot clients). Those clients must not be used for user tokens manipulations. The pilot clients will eventually carry the VO/group claims in them. BUT the issuer claim will be the same for all the VO's. The VO/group claims might be analysed on the sites in special plugins in order to identify the user's VO/group and perform an appropriate mapping to local accounts. This means that eventually it will be enough to have a single DIRAC client to run pilots for all the VOs. This is not yet the case, for the moment it is necessary to have a separate client per VO. 

So, the bottom line is that in the EGI/Check-In provider we will have to have at least 2 configured clients. By the way, as far as I can see in the ATLAS and CMS cases, they have separate user IDs for pilot submission, for example, for SAM/ETF tests: [topology](https://topology.opensciencegrid.org/collaborations/osg-scitokens-mapfile.conf). So, if even they will start using clients, they will have different clients for different purposes.
That's why I think we have to foresee the possibility to define multiple clients per identity provider. This should be done in a way that if a single client is necessary, no extra configuration is needed. I have coded this in the EGI dev service and these are just several extra lines of code. I insist that this possibility is included. 

I suggest to implement it in the following way. The TokenManagerClient/TokenManagerService getToken() interface is using IdP object client_credentials if the username is "client". The IdPFactory.getIdentityProvider() interface is checking kwargs["client"] value and, if present, uses it as the clientName: if it is given and not None, the IdProvider configuration dictionary is updated by the contents of the /Resources/IdProviders/<IdProvider>/Clients/<clientName> section. In the particular client section client_id/client_secret are overriding the default values in the  /Resources/IdProviders/<IdProvider>/ section. Finally, the TokenManagerClient/Service instantiate the IdP object with the client key argument; its value is simply equal to the pilot user group, e.g. biomed_pilot. If there is now  /Resources/IdProviders/<IdProvider>/Clients/<clientName> section, the defaults from /Resources/IdProviders/<IdProvider> are taken.  So, this is a minimal implementation and for cases where it will be possible to use a single client for all purposes, no extra configuration will be needed. I have a working code of this working, nothing really special but it differs much from this PR which includes other code refactorings. So, we have to see how this can be merged if we agree.

Reply:
@aldbr: We can remove the PilotDN option indeed but in many installations it is used in the configurations. So, I would suggest to keep it just in order not to create a mess for various services. Or at least it should be well announced in the release notes and in the meetings